src/Orleans.CodeGenerator/FieldIdAssignmentHelper.cs

```diff
@@ -0,0 +1,207 @@
+
+    private IEnumerable<ISymbol> GetMembers(INamedTypeSymbol symbol)
+    {
+        foreach (var member in symbol.GetMembers().OrderBy(m => m.MetadataName))
+        {
+            if (member.IsStatic || member.IsAbstract)
+            {
+                continue;
+            }
+
+            if (member is not (IFieldSymbol or IPropertySymbol))
+            {
+                continue;
+            }
+
+            if (member.HasAttribute(_libraryTypes.NonSerializedAttribute))
+            {
+                continue;
+            }
+
+            yield return member;
+        }
+    }
+
+    private bool ExtractFieldIdAnnotations()
+    {
+        foreach (var member in _memberSymbols)
+        {
+            if (member is IPropertySymbol prop)
+            {
+                var id = CodeGenerator.GetId(_libraryTypes, prop);
+
+                if (id.HasValue)
+                {
+                    _symbols[member] = (id.Value, false);
+                }
+            }
+
+            if (member is IFieldSymbol field)
+            {
+                var id = CodeGenerator.GetId(_libraryTypes, field);
+                var isConstructorParameter = false;
+
+                if (!id.HasValue)
+                {
+                    var property = GetMatchingProperty(field);
+                    if (property is null)
+                    {
+                        continue;
+                    }
+
+                    id = CodeGenerator.GetId(_libraryTypes, property);
+                    if (!id.HasValue)
+                    {
+                        var constructorParameter = _constructorParameters.FirstOrDefault(x => x.Name.Equals(property.Name, StringComparison.OrdinalIgnoreCase));
+                        if (constructorParameter is not null)
+                        {
+                            id = (ushort)_constructorParameters.IndexOf(constructorParameter);
+                            isConstructorParameter = true;
+                        }
+                    }
+                }
+
+                if (id.HasValue)
+                {
+                    _symbols[member] = (id.Value, isConstructorParameter);
+                }
+            }
+        }
+        return true;
+    }
+
+    private IPropertySymbol? GetMatchingProperty(IFieldSymbol field)
+    {
+        var propertyName = Regex.Match(field.Name, "^<([^>]+)>.*$");
+        if (!propertyName.Success)
+        {
+            return null;
+        }
+
+        var name = propertyName.Groups[1].Value;
+        var candidates = _memberSymbols.OfType<IPropertySymbol>()
+            .Where(property => string.Equals(name, property.Name, StringComparison.Ordinal)
+                               && SymbolEqualityComparer.Default.Equals(field.Type, property.Type)).ToArray();
+        return candidates.Length == 1 ? candidates[0] : null;
+    }
+
+    private IFieldSymbol? GetMatchingField(IPropertySymbol property)
```

Comment:
Any reason not to put these in a common place? They are mostly duplicates of existing methods

Reply:
The main reason is that they operate on a pre-filtered list of members (created in the `FieldIdAssignmentHelper` ctor). The performance gain is likely small, though. 

I've moved the methods to PropertyUtility and added overloads for passing in a pre-filtered member list, which should be a bit cleaner.